source/dofs/dof_tools_sparsity.cc

```diff
@@ -222,23 +225,62 @@ namespace DoFTools
+    // and dofs on a cell owned by a different processor.
+    constexpr int dim      = DoFHandlerType::dimension;
+    constexpr int spacedim = DoFHandlerType::space_dimension;
+    if (dynamic_cast<const parallel::Triangulation<dim, spacedim> *>(
+          &dof_row.get_triangulation()) != nullptr ||
+        dynamic_cast<const parallel::Triangulation<dim, spacedim> *>(
+          &dof_col.get_triangulation()) != nullptr)
+      {
+        Assert(&dof_row.get_triangulation() == &dof_col.get_triangulation(),
+               ExcMessage("This function can only be used with with parallel "
+                          "Triangulations when the Triangulations are equal."));
+      }
 
-    const std::list<std::pair<typename DoFHandlerType::cell_iterator,
-                              typename DoFHandlerType::cell_iterator>>
-      cell_list = GridTools::get_finest_common_cells(dof_row, dof_col);
+    // TODO: Looks like wasteful memory management here
 
+    using cell_iterator = typename DoFHandlerType::cell_iterator;
+    std::list<std::pair<cell_iterator, cell_iterator>> cell_list =
+      GridTools::get_finest_common_cells(dof_row, dof_col);
+
+#ifdef DEAL_II_WITH_MPI
+    // get_finest_common_cells returns all cells (locally owned and otherwise)
+    // for shared::Tria, but we don't want to assemble on cells that are not
+    // locally owned so remove them
+    if (dynamic_cast<const parallel::shared::Triangulation<dim, spacedim> *>(
+          &dof_row.get_triangulation()) != nullptr ||
+        dynamic_cast<const parallel::shared::Triangulation<dim, spacedim> *>(
+          &dof_col.get_triangulation()) != nullptr)
+      {
+        const types::subdomain_id this_subdomain_id =
+          dof_row.get_triangulation().locally_owned_subdomain();
+        Assert(this_subdomain_id ==
+                 dof_col.get_triangulation().locally_owned_subdomain(),
+               ExcInternalError());
+        cell_list.erase(
+          std::remove_if(
+            cell_list.begin(),
+            cell_list.end(),
+            [=](const std::pair<cell_iterator, cell_iterator> &pair) {
+              return pair.first->subdomain_id() != this_subdomain_id ||
+                     pair.second->subdomain_id() != this_subdomain_id;
+            }),
+          cell_list.end());
+      }
+#endif
 
-    typename std::list<std::pair<typename DoFHandlerType::cell_iterator,
-                                 typename DoFHandlerType::cell_iterator>>::
-      const_iterator cell_iter = cell_list.begin();
+    typename std::list<std::pair<cell_iterator, cell_iterator>>::const_iterator
+      cell_iter = cell_list.begin();
 
     for (; cell_iter != cell_list.end(); ++cell_iter)
       {
-        const typename DoFHandlerType::cell_iterator cell_row =
-          cell_iter->first;
-        const typename DoFHandlerType::cell_iterator cell_col =
-          cell_iter->second;
+        const cell_iterator cell_row = cell_iter->first;
```

Comment:
Can't we use a range-based for loop here?

Reply:
We can. I converted all of the typedefs but didn't convert the loop. I'll fix it.